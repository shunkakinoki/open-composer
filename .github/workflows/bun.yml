name: Bun
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
concurrency:
  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.sha }}
  cancel-in-progress: true
jobs:
  bun-build:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Build
        run: bun run build
  bun-lint:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Check
        run: bun run check
      - name: Format
        run: bun run format
      - name: Lint
        run: bun run lint
      - name: Type Check
        run: bun run type-check
  bun-test:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Test
        run: bun run test
  bun-check:
    if: always()
    needs:
      - bun-build
      - bun-lint
      - bun-test
    runs-on: ubuntu-latest
    timeout-minutes: 3
    steps:
      - name: Alls Green
        uses: re-actors/alls-green@release/v1
        with:
          jobs: ${{ toJSON(needs) }}
