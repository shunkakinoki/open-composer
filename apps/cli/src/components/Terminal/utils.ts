/**
 * Terminal utility functions
 * Inspired by termact: https://github.com/MasterGordon/termact/blob/main/src/utils.ts
 */

import type { TerminalStyle } from "./types.js";
import {
  bg,
  blink,
  bold,
  doubleUnderline,
  fg,
  hidden,
  inverse,
  italic,
  strikethrough,
  underline,
} from "./ansi.js";

export const applyStyle = (style: TerminalStyle): string => {
  let result = "";
  if (style.fg) result += fg(style.fg);
  if (style.bg) result += bg(style.bg);
  if (style.bold) result += bold();
  if (style.italic) result += italic();
  if (style.underline) result += underline();
  if (style.blink) result += blink();
  if (style.reverse) result += inverse();
  if (style.hidden) result += hidden();
  if (style.strikethrough) result += strikethrough();
  if (style.doubleUnderline) result += doubleUnderline();
  return result;
};

export const wrapText = (text: string, width: number): string[] => {
  const lines: string[] = [];
  const words = text.split(" ");
  let currentLine = "";

  for (const word of words) {
    if (currentLine.length + word.length + 1 <= width) {
      currentLine += (currentLine ? " " : "") + word;
    } else {
      if (currentLine) lines.push(currentLine);
      currentLine = word;
    }
  }

  if (currentLine) lines.push(currentLine);
  return lines;
};

export const truncateText = (text: string, maxLength: number): string => {
  if (text.length <= maxLength) {
    return text;
  }
  return text.slice(0, maxLength - 3) + "...";
};

export const padText = (
  text: string,
  width: number,
  align: "left" | "center" | "right" = "left",
): string => {
  if (text.length >= width) {
    return text.slice(0, width);
  }

  const padding = width - text.length;

  switch (align) {
    case "center": {
      const leftPad = Math.floor(padding / 2);
      const rightPad = padding - leftPad;
      return " ".repeat(leftPad) + text + " ".repeat(rightPad);
    }
    case "right":
      return " ".repeat(padding) + text;
    default:
      return text + " ".repeat(padding);
  }
};

export const getNtscGrayscale = (r: number, g: number, b: number): number => {
  return r * 0.299 + g * 0.587 + b * 0.114;
};
