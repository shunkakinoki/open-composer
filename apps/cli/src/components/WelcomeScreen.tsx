import { TextAttributes } from "@opentui/core";
import type React from "react";
import { useState } from "react";
import { CLI_VERSION } from "../lib/version.js";
import { MainMenu, type MenuItem } from "./MainMenu.js";
import { StatusBar } from "./StatusBar.js";

interface WelcomeScreenProps {
  onCommandSelect?: (commandName: string) => void;
}

export const WelcomeScreen: React.FC<WelcomeScreenProps> = ({
  onCommandSelect,
}) => {
  
  
  const [selectedCommand, setSelectedCommand] = useState<string | null>(null);

  // Terminal dimensions are handled by the renderer in opentui
  // These are not needed as the renderer handles terminal size automatically

  // Define menu items based on available commands
  const menuItems: MenuItem[] = [
    {
      key: "sessions",
      label: "Sessions",
      description: "List all sessions",
      onSelect: () => {
        setSelectedCommand("sessions");
        onCommandSelect?.("sessions");
      },
    },
    {
      key: "run",
      label: "Run",
      description: "Run a command",
      onSelect: () => {
        setSelectedCommand("run");
        onCommandSelect?.("run");
      },
    },
    {
      key: "spawn",
      label: "Spawn",
      description: "Spawn a new agent",
      onSelect: () => {
        setSelectedCommand("spawn");
        onCommandSelect?.("spawn");
      },
    },
    {
      key: "status",
      label: "Status",
      description: "Show status",
      onSelect: () => {
        setSelectedCommand("status");
        onCommandSelect?.("status");
      },
    },
    {
      key: "stack",
      label: "Stack",
      description: "Manage git stack",
      onSelect: () => {
        setSelectedCommand("stack");
        onCommandSelect?.("stack");
      },
    },
    {
      key: "settings",
      label: "Settings",
      description: "Configure settings",
      onSelect: () => {
        setSelectedCommand("settings");
        onCommandSelect?.("settings");
      },
    },
    {
      key: "tui",
      label: "Full TUI",
      description: "Launch full TUI mode",
      onSelect: () => {
        setSelectedCommand("tui");
        onCommandSelect?.("tui");
      },
    },
  ];

  const handleExit = () => {
    process.exit(0);
  };

  return (
    <box
      flexDirection="column"
    >
      {/* Header */}
      <box
        borderStyle="double"
        borderColor="cyan"
        justifyContent="center"
        paddingY={1}
      >
        <text content={`🎼 Open Composer CLI v${CLI_VERSION}`} style={{ fg: "cyan", attributes: TextAttributes.BOLD }} />
      </box>

      {/* Welcome Message */}
      <box
        flexDirection="column"
        padding={2}
        borderStyle="single"
        borderColor="gray"
      >
        <text content="Welcome to Open Composer! 👋" style={{ fg: "white", attributes: TextAttributes.BOLD }} />
        <text content="An agent orchestration framework for building with AI" style={{ fg: "gray" }} />
      </box>

      {/* Main Content Area */}
      <box flexGrow={1} flexDirection="row">
        {/* Main Menu */}
        <box width="60%">
          <MainMenu items={menuItems} onExit={handleExit} />
        </box>

        {/* Quick Info Panel */}
        <box
          width="40%"
          flexDirection="column"
          padding={1}
          borderStyle="single"
          borderColor="gray"
        >
          <text content="ℹ️ Quick Info" style={{ fg: "yellow", attributes: TextAttributes.BOLD }} />
          <box marginTop={1} flexDirection="column">
            <text content="• Select a command to get started" style={{ fg: "gray" }} />
            <text content="• Use arrow keys or j/k to navigate" style={{ fg: "gray" }} />
            <text content="• Press number keys for quick access" style={{ fg: "gray" }} />
            <text content="• Press Enter to execute command" style={{ fg: "gray" }} />
            <text content="• Press q to quit" style={{ fg: "gray" }} />
          </box>

          {selectedCommand && (
            <box marginTop={2} flexDirection="column">
              <text content="Selected:" style={{ fg: "green", attributes: TextAttributes.BOLD }} />
              <text content={selectedCommand} style={{ fg: "cyan" }} />
            </box>
          )}

          <box marginTop={2} flexDirection="column">
            <text content="🚀 Get Started" style={{ fg: "magenta", attributes: TextAttributes.BOLD }} />
            <box marginTop={1} flexDirection="column">
              <text content="Run 'open-composer --help' for CLI usage" style={{ fg: "gray" }} />
              <text content="Visit docs for more information" style={{ fg: "gray" }} />
            </box>
          </box>
        </box>
      </box>

      {/* Status Bar */}
      <StatusBar status="Ready" />
    </box>
  );
};
