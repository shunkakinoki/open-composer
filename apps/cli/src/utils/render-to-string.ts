import { render as inkRender } from "ink";
import type { ReactElement } from "react";
import { EventEmitter } from "node:events";

/**
 * Minimal writable stream for capturing Ink output
 */
class CaptureStream extends EventEmitter {
  private output = "";

  write(chunk: string): void {
    this.output += chunk;
  }

  getOutput(): string {
    return this.output;
  }

  end(): void {
    // no-op
  }
}

/**
 * Renders a React component to a string using Ink
 * Useful for rendering UI components in Effect CLI commands
 */
export function renderToString(element: ReactElement): string {
  const stdout = new CaptureStream();
  const stderr = new CaptureStream();

  const instance = inkRender(element, {
    stdout: stdout as unknown as NodeJS.WriteStream,
    stderr: stderr as unknown as NodeJS.WriteStream,
    debug: false,
    exitOnCtrlC: false,
    patchConsole: false,
  });

  // Wait for a tick to let Ink render
  const output = stdout.getOutput();

  // Clean up
  instance.unmount();
  instance.cleanup();

  return output;
}
