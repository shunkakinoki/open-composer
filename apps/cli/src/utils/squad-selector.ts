import inquirer from "inquirer";
import type { LLMTask } from "@open-composer/agent-registry";
import { getAgentRegistry, getSquadLauncher } from "@open-composer/agent-registry";

/**
 * Start interactive squad selector
 */
export async function startSquadSelector(): Promise<void> {
  const registry = getAgentRegistry();
  const launcher = getSquadLauncher();

  console.log("\n🎮 Pokemon-Style Squad Manager\n");

  const { action } = await inquirer.prompt([
    {
      type: "list",
      name: "action",
      message: "What would you like to do?",
      choices: [
        { name: "📋 List all agents", value: "list" },
        { name: "✨ Create new squad", value: "create" },
        { name: "🚀 Quick squad (by task)", value: "quick" },
        { name: "📊 View statistics", value: "stats" },
        { name: "🚪 Exit", value: "exit" },
      ],
    },
  ]);

  switch (action) {
    case "list": {
      const agents = registry.getAllAgents();
      console.log(`\n📦 Found ${agents.length} agents:\n`);
      for (const agent of agents) {
        console.log(
          `  ${agent.pokemonAttributes.sprite} ${agent.name} (${agent.tier}) - ${agent.provider}`,
        );
      }
      break;
    }

    case "create": {
      console.log("\n🛠️  Create Squad - Not implemented in CLI yet");
      console.log("   Use the interactive UI components for full squad creation");
      break;
    }

    case "quick": {
      const { task } = await inquirer.prompt([
        {
          type: "list",
          name: "task",
          message: "Select task specialization:",
          choices: [
            { name: "🧭 Planner", value: "planner" },
            { name: "⚡ Programmer", value: "programmer" },
            { name: "🔍 Reviewer", value: "reviewer" },
            { name: "🌊 Router", value: "router" },
            { name: "🌿 Summarizer", value: "summarizer" },
          ],
        },
      ]);

      const team = launcher.createQuickSquad(task as LLMTask);
      console.log(`\n✅ Created squad: ${team.name}`);
      console.log(`   Agents: ${team.agents.map((a) => a.name).join(", ")}`);
      break;
    }

    case "stats": {
      const stats = registry.getStats();
      console.log("\n📊 Registry Statistics:");
      console.log(`   Total Agents: ${stats.totalAgents}`);
      console.log(`   Total Squads: ${stats.totalTeams}`);
      console.log(`   Starter: ${stats.tierCounts.starter || 0}`);
      console.log(`   Evolved: ${stats.tierCounts.evolved || 0}`);
      console.log(`   Legendary: ${stats.tierCounts.legendary || 0}`);
      break;
    }

    case "exit":
      console.log("\n👋 Goodbye!\n");
      return;
  }

  // Loop back to main menu
  await startSquadSelector();
}
