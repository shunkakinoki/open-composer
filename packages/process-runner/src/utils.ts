import * as Effect from "effect/Effect";
import {
  MAX_COMMAND_LENGTH,
  MAX_RUN_NAME_LENGTH,
  RUN_NAME_PATTERN,
} from "./constants.js";
import { ProcessRunnerError } from "./types.js";

// =============================================================================
// VALIDATION UTILITIES
// =============================================================================

export function validateRunName(
  runName: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!runName || typeof runName !== "string") {
    return Effect.fail(
      ProcessRunnerError("Run name must be a non-empty string"),
    );
  }

  if (runName.length > MAX_RUN_NAME_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Run name too long (max ${MAX_RUN_NAME_LENGTH} characters)`,
      ),
    );
  }

  if (!RUN_NAME_PATTERN.test(runName)) {
    return Effect.fail(
      ProcessRunnerError(
        "Run name can only contain letters, numbers, hyphens, and underscores",
      ),
    );
  }

  return Effect.succeed(runName);
}

export function validateCommand(
  command: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!command || typeof command !== "string") {
    return Effect.fail(
      ProcessRunnerError("Command must be a non-empty string"),
    );
  }

  if (command.length > MAX_COMMAND_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Command too long (max ${MAX_COMMAND_LENGTH} characters)`,
      ),
    );
  }

  // Basic command injection prevention - only reject null bytes
  if (command.includes("\x00")) {
    return Effect.fail(
      ProcessRunnerError("Command contains invalid characters"),
    );
  }

  return Effect.succeed(command);
}

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

export function withTimeout<T>(
  effect: Effect.Effect<T, ProcessRunnerError>,
  timeoutMs: number,
  timeoutMessage: string = "Operation timed out",
): Effect.Effect<T, ProcessRunnerError> {
  const timeoutEffect = Effect.async<T, ProcessRunnerError>((resume) => {
    const timeout = setTimeout(() => {
      resume(Effect.fail(ProcessRunnerError(timeoutMessage)));
    }, timeoutMs);
    return Effect.sync(() => clearTimeout(timeout));
  });

  return Effect.race(effect, timeoutEffect);
}
