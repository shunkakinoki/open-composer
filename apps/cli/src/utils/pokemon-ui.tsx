import type { AgentTeam, OpenComposerAgent } from "@open-composer/agent-registry";
import type { RegistryStats } from "@open-composer/analytics";
import { renderToString } from "./render-to-string.js";
import {
  AgentCard,
  AgentComparison,
  AgentList,
  EvolutionTree,
  RegistryStats as RegistryStatsComponent,
  TaskSpecializations,
  TeamDisplay,
} from "../components/PokemonUI.js";

/**
 * Pokemon-style UI renderer using React components
 * This is a utility wrapper that renders React components to strings for CLI output
 */
export class PokemonUI {
  /**
   * Render agent card
   */
  static renderAgentCard(agent: OpenComposerAgent): string {
    return renderToString(<AgentCard agent={agent} />);
  }

  /**
   * Render agent list
   */
  static renderAgentList(agents: OpenComposerAgent[]): string {
    return renderToString(<AgentList agents={agents} />);
  }

  /**
   * Render team display
   */
  static renderTeam(team: AgentTeam): string {
    return renderToString(<TeamDisplay team={team} />);
  }

  /**
   * Render evolution tree
   */
  static renderEvolutionTree(agent: OpenComposerAgent): string {
    return renderToString(<EvolutionTree agent={agent} />);
  }

  /**
   * Render task specializations
   */
  static renderTaskSpecializations(agent: OpenComposerAgent): string {
    return renderToString(<TaskSpecializations agent={agent} />);
  }

  /**
   * Render registry statistics
   */
  static renderStats(stats: RegistryStats): string {
    return renderToString(<RegistryStatsComponent stats={stats} />);
  }

  /**
   * Render agent comparison
   */
  static renderComparison(
    agent1: OpenComposerAgent,
    agent2: OpenComposerAgent,
  ): string {
    return renderToString(<AgentComparison agent1={agent1} agent2={agent2} />);
  }
}
