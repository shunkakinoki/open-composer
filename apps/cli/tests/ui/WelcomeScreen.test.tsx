import { expect, mock, test } from "bun:test";
import { WelcomeScreen } from "../../src/components/WelcomeScreen.js";
import { render } from "../utils.js";

// Mock the version module to ensure consistent snapshots
mock.module("../../src/lib/version.js", () => ({
  CLI_VERSION: "0.0.0-test",
}));

test("WelcomeScreen renders header", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  const output = lastFrame();

  // Check for key parts of the header (accounting for border characters)
  expect(output).toMatch(/Open.*Composer.*CLI/);
  cleanup();
});

test("WelcomeScreen renders welcome message", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  const output = lastFrame();

  // Check for key words from the welcome message
  expect(output).toMatch(/agent.*orchestration.*framework/i);
  cleanup();
});

test("WelcomeScreen renders main menu items", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  const output = lastFrame();

  // Check for some key menu items
  expect(output).toMatch(/sessions/i);
  expect(output).toMatch(/Spawn/i);
  expect(output).toMatch(/status/i);
  cleanup();
});

test("WelcomeScreen renders quick info panel", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  const output = lastFrame();

  // Check for key parts of quick info
  expect(output).toMatch(/Quick.*Info/i);
  expect(output).toMatch(/arrow.*keys/i);
  cleanup();
});

test("WelcomeScreen renders status bar", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  const output = lastFrame();

  // Status bar contains text (may have formatting)
  expect(output.length).toBeGreaterThan(100);
  cleanup();
});

test("WelcomeScreen calls onCommandSelect when provided", async () => {
  let selectedCommand = "";
  const handleCommandSelect = (command: string) => {
    selectedCommand = command;
  };

  const { cleanup } = await render(<WelcomeScreen onCommandSelect={handleCommandSelect} />);

  // The component is rendered successfully
  expect(selectedCommand).toBe("");
  cleanup();
});

test("WelcomeScreen matches snapshot", async () => {
  const { lastFrame, cleanup } = await render(<WelcomeScreen />);
  expect(lastFrame()).toMatchSnapshot();
  cleanup();
});
