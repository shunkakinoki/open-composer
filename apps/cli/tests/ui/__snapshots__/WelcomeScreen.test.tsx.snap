// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`WelcomeScreen matches snapshot 1`] = `
"╔══════════════════════════════════════════════════════════════════════════════╗
╚🎼═Open═Composer═CLI═v0.0.0-test══════════════════════════════════════════════╝
┌──────────────────────────────────────────────────────────────────────────────┐
│                                                                              │
│                                                                              │
│  Anlagenttorchestrationeframework for building with AI                       │
│                                                                              │
└──────────────────────────────────────────────────────────────────────────────┘
                                                ┌──────────────────────────────┐
 Main Menu                                      │                              │
 ┌────────────────────────────────────────────┐ │ ℹ️ Quick Info                 │
 └────────────────────────────────────────────┘ │ • Useearrowckeysnortj/keto   │
                                                │ •aPressenumber keys for      │
 Listiall sessions                              │ •uPresscEnter to execute     │
 Spawn aonewnagent                              │ •oPress q to quit            │
 Showustatus                                    │                              │
 Configuretsettings                             │                              │
 LaunchUfull TUI mode                           │ 🚀 Get Started               │
                                                │ Visitodocscforomore--help'   │
                                                │ informationge                │
 Use ↑↓/j/k to navigate, Enter to select, 'q'   └──────────────────────────────┘
┌──────────────────────────────────────────────────────────────────────────────┐
│claude-code                                                        ?forqhelp| │
└──────────────────────────────────────────────────────────────────────────────┘
"
`;
